{
 "cells": [
  {
   "cell_type": "code",
   "id": "initial_id",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Data Cleaning \n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from pathlib import Path\n",
    "import sys\n",
    "\n",
    "# src-Verzeichnis relativ vom aktuellen Notebook-Verzeichnis\n",
    "sys.path.append(str(Path().resolve().parent / \"src\"))\n",
    "\n",
    "from src.data_cleaning import (batch_clean_all_tsv,\n",
    "                               mark_outliers,\n",
    "                               visualize_outliers\n",
    "                               )\n",
    "from src.data_handling import find_tsv_files\n",
    "from src.data_normalization import batch_normalize_tpm\n",
    "\n",
    "# Normalisierung der Rohdaten bevor Outlier entfernt werden\n",
    "#data1_folder = \"../data/raw_data_bb/\"\n",
    "#output1_folder = \"../data/pre_normalized_data_bb/\"\n",
    "#gff_folder = \"../data/raw_data_bb\"\n",
    "\n",
    "#batch_normalize_tpm(data1_folder, output1_folder, gff_folder, visual=False)\n",
    "\n",
    "# Konfiguration der Ein- und Ausgabeverzeichnisse (Bold-Binders Daten)\n",
    "#data_folder = \"../data/pre_normalized_data_bb/\"\n",
    "data_folder = \"../data/raw_data_bb/\"\n",
    "output_folder = \"../data/cleaned_data_bb/\"\n",
    "\n",
    "# Konfiguration der Ein- und Ausgabeverzeichnisse (Bold-Binders Daten)\n",
    "# data_folder = \"../data/raw_data_all/\"\n",
    "# output_folder = \"../data/cleaned_data_all/\"\n",
    "\n",
    "batch_clean_all_tsv(data_folder, output_folder, method='iqr',\n",
    "                    visual=True)  # der Parameter visual steuert die Ausgabe der Visualisierung"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Normalization\n",
    "from src.data_normalization import batch_normalize_tpm\n",
    "\n",
    "data_folder = \"../data/cleaned_data_bb/\"\n",
    "output_folder = \"../data/normalized_data_bb/\"\n",
    "gff_folder = \"../data/raw_data_bb\"\n",
    "\n",
    "# # Konfiguration der Ein- und Ausgabeverzeichnisse (Aller Daten)\n",
    "# data_folder = \"../data/cleaned_data_all/\"\n",
    "# output_folder = \"../data/normalized_data_all/\"\n",
    "# gff_folder = \"../data/raw_data_all\"\n",
    "\n",
    "batch_normalize_tpm(data_folder, output_folder, gff_folder, phage_only=True,\n",
    "                    visual=True)  # der Parameter visual steuert die Ausgabe der Visualisierung"
   ],
   "id": "7a048181976d2442",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# sample based PCA\n",
    "from src.sample_pca import process_directory\n",
    "# Ausführung\n",
    "input_dir = \"../data/normalized_data_bb\"\n",
    "# Benutzerdefinierte Ausreißer (optional)\n",
    "# user_defined_outliers = [\"20_R3\", \"12_R1\"]\n",
    "# user_defined_outliers = None #automatiziert\n",
    "user_defined_outliers = []\n",
    "\n",
    "file_name = \"Wolfram\"\n",
    "process_directory(input_dir, user_outliers=user_defined_outliers, filename_filter=file_name)\n",
    "# process_directory(input_dir, user_outliers=user_defined_outliers)"
   ],
   "id": "78bd28a320713f03",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# US4-7 Labeling",
   "id": "9a0e79b82489e500"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Labeling\n",
    "from src.labeling import process_all_files, labeling_pipeline_means\n",
    "from src.visualization import (\n",
    "    plot_label_distribution_comparison,\n",
    "    plot_aggregated_distribution\n",
    ")\n",
    "\n",
    "# Konfiguration der Ein- und Ausgabeverzeichnisse (Bold-Binders Daten)\n",
    "data_folder = \"../data/normalized_data_bb/\"\n",
    "output_folder = \"../data/gene_labeling_bb/\"\n",
    "time_means_folder=\"../data/time_means/\"\n",
    "\n",
    "# # Konfiguration der Ein- und Ausgabeverzeichnisse (Aller Daten)\n",
    "# data_folder = \"../data/normalized_data_all/\"\n",
    "# output_folder = \"../data/gene_labeling_all/\"\n",
    "\n",
    "# Verarbeitung aller Dateien\n",
    "label_counts_std, label_counts_paper, _ = process_all_files(data_folder, output_folder)\n",
    "label_counts_means = labeling_pipeline_means(data_folder,time_means_folder, output_folder, \"gene_labels_means.tsv\")\n",
    "\n",
    "# Erstelle Visualisierungen\n",
    "fig1 = plot_label_distribution_comparison(label_counts_std, label_counts_paper, label_counts_means)\n",
    "\n",
    "fig2 = plot_aggregated_distribution(label_counts_std, label_counts_paper, label_counts_means)\n",
    "\n",
    "print(f\"Alle Ergebnisse wurden im Verzeichnis '{output_folder}' gespeichert.\")"
   ],
   "id": "f8b9380fe6b460c0",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "US8 Genome Maps",
   "id": "1da360e976a12ebe"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Genome Maps\n",
    "from src.genome_map import visualize_from_merged_tsv_by_geneid\n",
    "\n",
    "merged_tsv_path = \"../data/gene_labeling_bb/gene_labels_means.tsv\"\n",
    "gff_phage_dir = \"../data/raw_data_bb/phage_gff_files\"\n",
    "output_genome_maps_dir = \"../data/genome_maps\"\n",
    "\n",
    "visualize_from_merged_tsv_by_geneid(\n",
    "    merged_tsv_path,\n",
    "    gff_phage_dir,\n",
    "    output_genome_maps_dir\n",
    ")"
   ],
   "id": "75ae65e780a72a14",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "",
   "id": "4b73df027cc78a7c",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "US9 Feature Extraction",
   "id": "afb7c33c7f18a583"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "from pathlib import Path\n",
    "import sys\n",
    "\n",
    "# src-Verzeichnis relativ vom aktuellen Notebook-Verzeichnis\n",
    "sys.path.append(str(Path().resolve().parent / \"src\"))\n",
    "\n",
    "from src.feature_extraction import extract_features\n",
    "\n",
    "# Pfade definieren\n",
    "gff_dir = \"../data/raw_data_bb/phage_gff_files\"\n",
    "fasta_dir = \"../data/raw_data_bb\"\n",
    "#Auswahl des Labeling Verfahrens\n",
    "labels_file = \"../data/gene_labeling_bb/gene_labels_means.tsv\"\n",
    "output_file = \"../data/gene_features/gene_features.tsv\"\n",
    "\n",
    "# Stelle sicher, dass das Ausgabeverzeichnis existiert\n",
    "Path(output_file).parent.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "try:\n",
    "    # Feature Extraktion ausführen\n",
    "    results = extract_features(\n",
    "        gff_directory=gff_dir,\n",
    "        fasta_directory=fasta_dir,\n",
    "        labels_file=labels_file,\n",
    "        output_file=output_file\n",
    "    )\n",
    "\n",
    "    if results is not None and not results.empty:\n",
    "        # Erste paar Zeilen anzeigen\n",
    "        print(\"\\nErste Einträge der extrahierten Daten:\")\n",
    "        display(results.head())\n",
    "\n",
    "        # Statistiken anzeigen\n",
    "        print(\"\\nStatistiken:\")\n",
    "        print(f\"Anzahl der extrahierten Gene: {len(results)}\")\n",
    "        print(f\"Anzahl einzigartiger Phagen: {results['source_study'].nunique()}\")\n",
    "        print(f\"Verteilung der Labels:\\n{results['label'].value_counts()}\")\n",
    "    else:\n",
    "        print(\"Keine Daten wurden extrahiert.\")\n",
    "\n",
    "except Exception as e:\n",
    "    print(f\"Fehler bei der Feature-Extraktion: {str(e)}\")\n"
   ],
   "id": "fd8b4bcb57272a6c",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Feature-Matrix erzeugen, laden und visualisieren\n",
    "import pandas as pd\n",
    "import sys\n",
    "import os\n",
    "from pathlib import Path\n",
    "\n",
    "# src-Verzeichnis hinzufügen\n",
    "sys.path.append(str(Path().resolve().parent / \"src\"))\n",
    "\n",
    "# build_features.py einmalig ausführen (führt gesamte Verarbeitung aus)\n",
    "os.system(\"python ../src/build_features.py\")\n",
    "\n",
    "# Danach Matrix laden und visualisieren\n",
    "from src.visualization import visualize_feature_matrix\n",
    "\n",
    "df = pd.read_csv(\"../data/feature_matrix/feature_matrix.tsv\", sep=\"\\t\")\n",
    "visualize_feature_matrix(df)\n"
   ],
   "id": "db852060445d9356",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# US14 Trainings-/Test-Split auf Gesamtdatensatz\n",
    "# Split-Verhältnis: 80/20\n",
    "# Seed: 42\n",
    "from src.split import modified_find_tsv_file, load_tsv, split_stratified, print_label_distribution, save_splits_to_tsv, plot_label_distribution\n",
    "\n",
    "data_dir = \"../data/feature_matrix\"\n",
    "split_output_dir=\"../data/split_dataset_us14\"\n",
    "filepath = modified_find_tsv_file(data_dir)\n",
    "df = load_tsv(filepath)\n",
    "train_df, test_df = split_stratified(df)\n",
    "print_label_distribution(df, train_df, test_df)\n",
    "save_splits_to_tsv(train_df, test_df, split_output_dir)\n",
    "plot_label_distribution(df, train_df, test_df)"
   ],
   "id": "7393f2256b4a6995",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Daten vorverarbeiten und bereinigen, bevor der Testdatensatz erstellt wird für US15\n",
    "from pathlib import Path\n",
    "import sys\n",
    "\n",
    "# src-Verzeichnis relativ vom aktuellen Notebook-Verzeichnis\n",
    "sys.path.append(str(Path().resolve().parent / \"src\"))\n",
    "\n",
    "from src.new_test_data_u15.new_test_data_preprocessing import run_preprocessing_pipeline\n",
    "\n",
    "run_preprocessing_pipeline(\n",
    "    raw_data_dir=\"../data/new_test_data/raw_data_test/\",\n",
    "    cleaned_data_dir=\"../data/new_test_data/cleaned_data_test/\",\n",
    "    normalized_data_dir=\"../data/new_test_data/normalized_data_test/\",\n",
    "    time_means_folder = \"../data/new_test_data/time_means/\",\n",
    "    labeled_data_dir=\"../data/new_test_data/gene_labeling_test/\",\n",
    "    feature_output_path=\"../data/new_test_data/gene_features_test/gene_features.tsv\",\n",
    "    build_script_path=\"../src/new_test_data_u15/build_features_new_test_data.py\",\n",
    "    visualize=False\n",
    ")"
   ],
   "id": "a2cd0d1ef8a7338",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# US15: Definition einzelner Datensätze als Testsets\n",
    "# Test - Datensätze: Yang\n",
    "from src.new_test_data_u15.split_new_test_data import run_us15_split\n",
    "\n",
    "train_dataset_dir = \"../data/feature_matrix\"\n",
    "test_dataset_dir = \"../data/new_test_data/feature_matrix_test\"\n",
    "us15_dataset_dir = \"../data/new_test_data/split_dataset_us15\"\n",
    "key_column = 'id'\n",
    "label_column = 'label'\n",
    "\n",
    "run_us15_split(\n",
    "        train_dir = train_dataset_dir,\n",
    "        test_dir = test_dataset_dir,\n",
    "        output_dir = us15_dataset_dir,\n",
    "        key_column = key_column,\n",
    "        label_column = label_column\n",
    ")"
   ],
   "id": "1254ba23ded9c4b6",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "#### US16 Vergleich verschiedener Split-Strategien\n",
    "**Strategie 1**: Trainings-/Test-Split auf Gesamtdatensatz(US14)<br>\n",
    "    Organismen von Gesamtdatensatz: Pseudomonas_phage_LUZ19; Pseudomonas_phage_phiKZ; Enterobacteria_phage_T7; Yersinia_phage_phiR1-37; Clostridium_phage_JD032; Vibrio_phage_VP882; Enterobacteria_phage_T4<br>\n",
    "**Strategie 2**: Definition einzelner Datensätze als Testsets(US15)<br>\n",
    "    Organismen von Trainingsdatensatz: Pseudomonas_phage_LUZ19; Pseudomonas_phage_phiKZ; Enterobacteria_phage_T7; Yersinia_phage_phiR1-37; Clostridium_phage_JD032; Vibrio_phage_VP882; Enterobacteria_phage_T4<br>\n",
    "    Organismus von Testdatensatz: Acinetobacter_phage<br>\n",
    "das angewendete Modell: **Random Forest**\n",
    "- Im Skript 'src/compare_splits' wurde das Modell **Random Forest** in der Funktion 'train_and_predict' integriert. Das Skript ist so aufgebaut, dass interne Parameter des Random Forest bei Bedarf angepasst und erneut verwendet werden können.\n",
    "- Verwendete Klassifikationsmetriken: Accuracy, AUC, Precision, Recall und F1-score, Support. Zusätzlich wurden die Metriken Precision, Recall und F1-score auch für jede einzelne Klasse berechnet (early, middle, late). Die letzte Zeile zeigt die Differenz zwischen den Strategien, um relative Leistungsunterschiede zu zeigen.\n",
    "- Insgesamt zeigen sich keine großen Unterschiede zwischen den beiden Strategien. Strategy 2 (US15) zeigt insgesamt leicht bessere Klassifikationsmetriken, insbesondere bei AUC und F1-Score. Jedoch bietet Strategy 1 (US14) aufgrund der größeren Testmenge stabilere und zuverlässigere Ergebnisse."
   ],
   "id": "974acda6c4fa5d3c"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "from src.compare_splits import (\n",
    "    run_comparison,\n",
    "    plot_heatmap\n",
    ")\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "import pandas as pd\n",
    "\n",
    "label_encoder = LabelEncoder()\n",
    "\n",
    "results_df = run_comparison(\n",
    "    \"../data/split_dataset_us14/train_set.tsv\",\n",
    "    \"../data/split_dataset_us14/test_set.tsv\",\n",
    "    \"../data/new_test_data/split_dataset_us15/train_set.tsv\",\n",
    "    \"../data/new_test_data/split_dataset_us15/test_set.tsv\",\n",
    "    label_encoder\n",
    ")\n",
    "\n",
    "pd.set_option(\"display.precision\", 4)\n",
    "print(\"\\nVergleich der Metriken:\")\n",
    "print(results_df)\n",
    "\n",
    "plot_heatmap(results_df)"
   ],
   "id": "3715779222fc04d8",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "#### US17: Vergleich mehrerer Klassifikator-Algorithmen\n",
    "Da die bisherige Feature-Matrix ganz anders aufgebaut ist als die Embeddings aus US12, haben wir drei Modelle (SVM, Random Forest und Decision Tree) ohne Embeddings, sondern ausschließlich mit den von Hand erstellten Features trainiert.\n",
    "\n",
    "Hauptmodell: MLP<br>\n",
    "Benchmarkmodelle\n",
    "- Modell 1: Decision Tree\n",
    "- Modell 2: Random Forest\n",
    "- Modell 3: SVM\n",
    "\n",
    "Dataset-Split-Strategie: Trainings-/Test-Split(80/20) auf Gesamtdatensatz(US14)"
   ],
   "id": "a1139c6a9b2b775c"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-07-25T15:17:59.538765Z",
     "start_time": "2025-07-25T15:17:53.860521Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Decision Tree\n",
    "# final model script: src/decision_tree.py\n",
    "# grid search script for Decision Tree: src/grid_search_decision_tree.py\n",
    "from importlib.util import spec_from_file_location, module_from_spec\n",
    "import joblib\n",
    "import sys\n",
    "\n",
    "def load_model_module(path, module_name):\n",
    "    \"\"\"\n",
    "    Lädt ein Python-Modul dynamisch aus dem gegebenen Pfad mit dem gewünschten Modulnamen.\n",
    "    \"\"\"\n",
    "    spec = spec_from_file_location(module_name, path)\n",
    "    module = module_from_spec(spec)\n",
    "    sys.modules[module_name] = module\n",
    "    spec.loader.exec_module(module)\n",
    "    return module\n",
    "\n",
    "# Module laden\n",
    "dt_module = load_model_module(\"../src/decision_tree.py\", \"decision_tree\")\n",
    "svc_module = load_model_module(\"../src/support_vector_machine.py\", \"svc\")\n",
    "\n",
    "# Modellpfad definieren\n",
    "try:\n",
    "    model_path = \"../models/final_svm_model.pkl\"\n",
    "except Exception as e:\n",
    "    print(f\"modell nicht gefunden {str(e)}\")\n",
    "\n",
    "# Vortrainiertes Modell laden\n",
    "final_svm_model = joblib.load(model_path)\n",
    "print(type(final_svm_model))             # Sollte <class 'sklearn.svm._classes.SVC'> sein\n",
    "print(hasattr(final_svm_model, \"support_\"))  # Sollte True sein, wenn das Modell trainiert wurde\n",
    "\n",
    "print(\"Decision Tree Modell: \")\n",
    "# Trainings- und Testdaten laden, Modell trainieren und auswerten\n",
    "dt_module.train_and_evaluate(\n",
    "    dt_module.final_model,\n",
    "    \"../data/split_dataset_us14/train_set.tsv\",\n",
    "    \"../data/split_dataset_us14/test_set.tsv\"\n",
    ")\n",
    "\n",
    "print(\"Support Vector Machine Modell:\")\n",
    "#svc_module.train_and_evaluate(\n",
    "#    final_svm_model,\n",
    "#    \"../data/split_dataset_us14/train_set.tsv\",\n",
    "#    \"../data/split_dataset_us14/test_set.tsv\"\n",
    "#)\n",
    "svc_module.model_predict(final_svm_model, \"../data/split_dataset_us14/test_set.tsv\")"
   ],
   "id": "95f9cade8aa9af38",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'sklearn.svm._classes.SVC'>\n",
      "True\n",
      "Decision Tree Modell: \n",
      "Klassifikationsbericht (Testdaten):\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "       early       0.44      0.49      0.46       104\n",
      "        late       0.41      0.42      0.42        83\n",
      "      middle       0.24      0.19      0.21        64\n",
      "\n",
      "    accuracy                           0.39       251\n",
      "   macro avg       0.36      0.37      0.36       251\n",
      "weighted avg       0.38      0.39      0.38       251\n",
      "\n",
      "Support Vector Machine Modell:\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "       early       0.56      0.48      0.52       104\n",
      "        late       0.41      0.37      0.39        83\n",
      "      middle       0.50      0.67      0.57        64\n",
      "\n",
      "    accuracy                           0.49       251\n",
      "   macro avg       0.49      0.51      0.49       251\n",
      "weighted avg       0.49      0.49      0.49       251\n",
      "\n"
     ]
    }
   ],
   "execution_count": 3
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
